Rich editor for pastes

tachyondecay opened the pull request:
Test at: https://qp.tachyondecay.net/

Works for me in the most recent versions of Firefox and Chrome. Should be tested still in Safari and Edge (and maybe IE10?).

### Changes
* `RICH_EDITOR_ENABLED` config flag toggles all these features. Default: `false` (no rich editor, functions as it used to)
* Degrades gracefully when JavaScript disabled
* Code highlights every time a user stops typing for at least 1 second
* Highlighting only happens when a language is selected. I did this because Pygments would very often guess wrong (usually it would guess MIME lexer), and the result is not great. Selecting a language triggers highlighting.
* Pasting code will try to auto-detect indentation style (and width, if spaces are being used)
* Indentation style can be changed on the fly and code will be re-indented
* Pressing enter should insert a new line with current indentation level (sometimes buggy)

### Bugs
Caret position is liable to jump around at times, especially after highlighting has been applied, or when doing newlines. I don't think it's possible to completely eliminate buggy behaviour (this is a limitation of contenteditable + browser compatibility issues + modifying the HTML). This is the tradeoff for not simply using a heavier, pre-existing rich editor library. My hope is that I've eliminated the most annoying bugs so that the editor is usable.

### Minor changes to other files
* Pygments highlighting calls abstracted out to `util.py` now that they are involved in both editing and viewing. This does not change *how* Pygments is called, just simplifies the process
* Some changes in the CSS were required, e.g., the line-numbering now uses flexbox. ### Visually the differences are minor.
* Updated README with new config value; didn't update it or about.md with any mention of the rich editor. I'll leave that to you when you decide to release it.

carc1n0gen commented:
I definitely ran in to the cursor jumping around my very first test.  Was just typing out a C hello world.  I know you mentioned it, but it seems quite bad that I couldn't just type a small program.

![Jun-12-2020 07-56-39](https://user-images.githubusercontent.com/8248986/84500644-f7b85480-ac82-11ea-88f1-fb885017ca0c.gif)

Also something specifically with pasting a C snippet.  I think it is related to the use of `<` and `>` causes some code to be deleted when highlighting is triggered I'm guessing.

![Jun-12-2020 07-57-22](https://user-images.githubusercontent.com/8248986/84500794-3c43f000-ac83-11ea-9b83-7c824568e2c3.gif)

Another thing in that last gif is at the end.  If I Cmd+a (or Ctr+a) to highlight all, it unhighlights itself just a moment after.

But I won't let these stop the PR from being merged, since the feature can be left turned off while it is improved.  Going to begin reviewing the changed now.

carc1n0gen commented on app/blueprints/edit.py:
```diff
@@ -93,14 +93,24 @@ def edit():
+
+@edit_bp.route('/highlight', methods=['POST'])
+def live_highlight():
+    text = request.form.get('text')
```
I'd like to use wtforms here so that we have csrf protection.  In case of the unlikely event that someone decides to use this endpoint as a free highlighting api.

tachyondecay replied:
Done.

carc1n0gen commented on static-src/js/editor.js:
```diff
@@ -0,0 +1,290 @@
+            }
+
+            if(e.key == "Tab") {
+                e.preventDefault();
```
I'd be interested in adding Shift+Tab for un-indent.

carc1n0gen commented on static-src/js/editor.js:
```diff
@@ -0,0 +1,290 @@
+        // Autoformat code when indentation style changes
+        this.indentControl.addEventListener('change', e => {
+            let old_size = this.indentation.size;
+            let type, size;
```
If these don't get resigned to, it could be changed to `const [type, size] = this.indentControl.value.split('-');`

tachyondecay replied:
Done.

carc1n0gen commented on static-src/js/editor.js:
```diff
@@ -0,0 +1,290 @@
+            'type': 'spaces'
+        };
+
+        const matches = this.editor.innerText.match(/^[ \t]+/gm);
```
If we only ever use the first index, could change to `const [match] = this.editor.innerText.match(/^[ \t]+/gm);`

tachyondecay replied:
Done.

carc1n0gen commented on static-src/js/editor.js:
```diff
@@ -0,0 +1,290 @@
+
+    // Dispatch a request to app for syntax highlighting
+    async highlightRequest() {
+        const data = new FormData();
```
If we do add wtforms (and by effect, csrf protection), we'll need a way to put the token in this request.

tachyondecay replied:
Done. This will now return a JSON object of the form errors in the event of any errors, so in the future someone could turn that into nice toasts for the user if we care to.

carc1n0gen replied:
That's actually perfect.  I've already got a toast that only shows when the paste is longer than the configured max.  I can hook in to that later.

carc1n0gen commented on templates/edit/edit.html:
```diff
@@ -1,16 +1,35 @@
         </div>
         {% endif %}
+        <div class="select-wrapper indent-control rich-disabled">
+            <label for="indent">Indentation:</label>
```
Can we either add `sr-only` to the label, or instead of a label element, use `aria-label="..."` on the select element?

tachyondecay replied:
Done.

tachyondecay commented:
Fixed the Ctrl+A problem and pasting code with angle brackets. I had added a "fix" for doing some newline inserts that broke my previous fix for angle brackets being swallowed up, oops.

> EDIT: Oh also the placeholder is missing now, probably due to the content editable box. But very minor.

You know I got you. It's back, grabs the text from the textarea's placeholder, removes it on focus and replaces it on blur (if editor is empty).

### Indenting (Tab and Shift Tab)
Per our Discord DM, we can get some block-level indentation happening. I just need to do a bit of a rewrite on the indentation-insertion handling. This *may* involve rewriting how newlines get handled as well. Basically I might look for a more "canonical" way to have the editor track newlines relative to the current cursor position. This is "easy" if you only have text nodes, but the fact that we have non-text nodes mixed in from the highlighting makes it more challenging. This is where I start copy-pasting code from StackOverflow, and I really hate doing that unless I know exactly how it works. I'm comfortable with the "save caret position" stuff I used, but some of the more "complete" solutions are beasts.

### Cursor troubles
I think the most egregious cursor-jumping is happening because of this chain of events:

1. User stops typing for 1 second, triggering async highlight request.
1. User resumes typing before highlight request returns.
1. Highlight request returns and restores the position of the cursor (this is required when completely replacing the innerHTML).
1. Since the user's cursor has moved since then, this is a "jump."

I think the solution is to abort a pending highlight request if the user starts typing again. I looked around for an "easy" way to do this with the Fetch API but couldn't find anything I could implement in about 10 minutes, so I shelved it in favour of showing you what I had so far. But I think I can either:

- Implement a true "abort" of the Fetch request
- Somehow mark pending requests as "stale" and just ignore their response when it comes back (this requires a little more logic to track state)

I'm happy to keep working on this, whether or not you merge the current commits. We could also merges this into a separate `rich-editor` branch on your repo if you want to keep these siloed and even work on it yourself until you're ready to merge them into the production version of the app.

carc1n0gen commented:
Merging as is.  The outstanding issues can be worked on when time permits.
